### DISCUSSION TITLE
How to compute the complexity of a function inside the custom objective function?

### DISCUSSION THREAD

**Discussion by unary-code** [MilesCranmer/PySR#380] [Q&A] [answered]
This is a question more for the SymbolicRegression julia package than the PySR package.

I am using the PySR package, and I am using a custom objective function.

1. Inside the custom objective function, I want to get the complexity of whatever tree is passed into the custom objective function. How do I do this? Without recursively searching the left child and right child of the tree, which I feel would take too long.

2. Also, is there a way to force the model to only use up to 2 exp() (e^power) operators in the functions it generates?

**Comment by MilesCranmer** [comment, marked as answer]
1. You can do it with `compute_complexity(tree, options)`
2. Yes, see `nested_constraints`: https://astroautomata.com/PySR/api/#working-with-complexities. If you mean limiting the total number in the expression, you will have to do that manually in the loss function.

For example, if we assume that `exp` is the 3rd operator in the list of unary operators, this would be:

```julia
num_exps = count(node -> node.degree == 1 && node.op == 3, tree)
```

Then you can return some large number if `num_exps > 2`. You can also do it like this if this is more readable:

```julia
num_exps = 0
for node in tree
    if node.degree == 1 && node.op == 3
        num_exps += 1
    end
end
```

**Reply by unary-code** [reply]
Oh okay, thanks for your answer.

I noticed that the "compute_complexity" function is in one of the example codes on the SymbolicRegression jl documentation website (https://astroautomata.com/SymbolicRegression.jl/stable/examples/), but there is no documentation for it. Not a issue, because this function seems easy to understand.

"nested_constraints" is what I want to use now, but the counting of the total number of "exp"'s in the expression might be useful later on. Thanks for explaining!

**Reply by MilesCranmer** [reply]
Oops, good point. It does have a docstring: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/0f47c6baf783b436ccecd5e635f692516c92d963/src/Complexity.jl#L10-L24

but it seems like it doesn't show up on the website due to not being included here: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/master/docs/src/api.md
